src/controller/cmcd-controller.ts

```diff
@@ -343,6 +352,67 @@ export default class CMCDController implements ComponentAPI {
     return info.len * 1000;
   }
 
+  /**
+   * Create a playlist loader
+   *
+   * @returns {PlaylistLoaderConstructor | undefined}
+   */
+  private createPlaylistLoader(): PlaylistLoaderConstructor | undefined {
+    const { pLoader } = this.config;
+
+    if (!this.config.cmcdEnabled) {
+      return pLoader;
+    }
+
+    const apply = this.applyPlaylistData;
+    const ctor = pLoader || (this.config.loader as PlaylistLoaderConstructor);
+
+    // @ts-ignore
+    return function (config: HlsConfig) {
+      const loader = new ctor(config);
+      this.abort = () => loader.abort();
+      this.destroy = () => loader.destroy();
+      this.load = (
+        context: PlaylistLoaderContext,
+        config: LoaderConfiguration,
+        callbacks: LoaderCallbacks<PlaylistLoaderContext>
+      ) => {
+        apply(context);
+        loader.load(context, config, callbacks);
+      };
+    };
+  }
+
+  /**
+   * Create a playlist loader
+   *
+   * @returns {FragmentLoaderConstructor | undefined}
+   */
+  private createFragmentLoader(): FragmentLoaderConstructor | undefined {
+    const { fLoader } = this.config;
+
+    if (!this.config.cmcdEnabled) {
+      return fLoader;
+    }
+
+    const apply = this.applyFragmentData;
+    const ctor = fLoader || (this.config.loader as FragmentLoaderConstructor);
+
+    // @ts-ignore
+    return function (config: HlsConfig) {
+      const loader = new ctor(config);
+      this.abort = () => loader.abort();
+      this.destroy = () => loader.destroy();
+      this.load = (context, config, callbacks) => {
+        apply(context);
+        loader.load(context, config, callbacks);
+      };
+      Object.defineProperty(this, 'stats', {
+        get: () => loader.stats,
+      });
+    };
+  }
+
```

Comment:
This is the main part of the refactor. When the `CMCDController` is instantiated, it will compose new loaders based on the `HlsConfig` using these functions. The rest of the changes in this PR deal with removing the previous methodology of transforming the context.

Reply:
This indeed looks much cleaner.

The one thing which might deserve some discussions is the fact that the new component is implicitly overriding some config keys (pLoader / fLoader)
```
    this.config.pLoader = this.createPlaylistLoader();
    this.config.fLoader = this.createFragmentLoader();
```
This override seems to also happen if CMCD is not enabled.

Thinking loud, but If I am not mistaken, cmcd-controller.ts is now standalone controller , listening to events, ... 
I am wondering whether it might make sense to have cmcd-controller.ts within hls.js repo, but not bundled in the main dist. so that this overriding is only done on purpose, when CMDController is being used explicitly.

in order to use CMCD you would then have to do something like

hls = new Hls()
cmcdController = new CMCDController(hls)

we would also need to destroy it properly on hls.destroy()

Reply:
I've updated `CMCDController` to only override the loaders and add listeners when `config.cmcdEnabled` is `true`.

Reply:
@mangui I'm going to merge this into the `videodev/hls.js` PR. Can you please add your bundling comment to the original PR?